### PR TITLE
Implement many minor deprecations in v4.0

### PR DESCRIPTION
**Pull request by Zac-HD**
There are dozens of relatively minor deprecations scattered through our code - from changes to settings, to tighter validation, or new checks for problematic conditions - saying something like "X is bad.  In a future version, [X will be an error | Hypothesis will Y]."

This is the PR that does the future thing.  I strongly recommend reviewing each commit individually, as they're self-contained and seeing how the edits are grouped is pretty useful.

### CODE REVIEW

File: src/hypothesis/strategies.py [hunk 1]
```diff
@@ -461,9 +462,35 @@ def sampled_from(elements):
     1 values with 10, and sampled_from((1, 10)) will shrink by trying to
     replace 10 values with 1.
     """
-    from hypothesis.searchstrategy.misc import SampledFromStrategy
-    from hypothesis.internal.conjecture.utils import check_sample
-    values = check_sample(elements)
+    try:
```
Comment (alexwlchan) [inline src/hypothesis/strategies.py]: I’d put this in `hypothesis.internal.validation`, which is where a lot of other strategy validation logic lives (e.g. `check_valid_bound()` in `integers()`).

The original `check_samples()` was arguably in the wrong place too, but that’s less important now.

File: src/hypothesis/strategies.py [hunk 2]
```diff
@@ -461,9 +462,35 @@ def sampled_from(elements):
...
+                raise InvalidArgument(
+                    'Only one-dimensional arrays are supported for sampling, '
+                    'and the given value has %s dimensions (shape %s).  '
+                    'This array would give samples of array slices '
+                    'instead of elements!  Use np.ravel(values) to convert '
+                    'to a one-dimensional array, or tuple(values) if you '
+                    'want to sample slices.' % (elements.ndim, elements.shape)
+                )
+            elements = tuple(elements)
+    except ImportError:  # pragma: no cover
```
Comment (alexwlchan) [inline src/hypothesis/strategies.py]: minor: I’m surprised we can’t get coverage on this line.

File: src/hypothesis/searchstrategy/strategies.py [hunk 3]
```diff
@@ -238,7 +237,7 @@ def example(self, random=None):
         context = _current_build_context.value
         if context is not None:
             if context.data is not None and context.data.depth > 0:
-                note_deprecation(
+                raise InvalidState(
                     'Using example() inside a strategy definition is a bad '
                     'idea. It will become an error in a future version of '
```
Comment (alexwlchan) [inline src/hypothesis/searchstrategy/strategies.py]: You need to edit this message – it won’t become an error, it is an error!

File: src/hypothesis/searchstrategy/strategies.py [hunk 4]
```diff
@@ -252,7 +251,7 @@ def example(self, random=None):
                     '#composite-strategies for more details.'
                 )
             else:
-                note_deprecation(
+                raise InvalidState(
                     'Using example() inside a test function is a bad '
                     'idea. It will become an error in a future version of '
```
Comment (alexwlchan) [inline src/hypothesis/searchstrategy/strategies.py]: Ditto above.

File: src/hypothesis/database.py [hunk 5]
```diff
@@ -106,7 +90,7 @@ def __init__(self):
         self.data = {}
 
     def __repr__(self):
-        return 'InMemoryExampleDatabase(%r)' % (self.data,)
+        return '%s(%r)' % (self.__class__.__name__, self.data)
```
Comment (alexwlchan) [inline src/hypothesis/database.py]: 👍

**Comment by Zac-HD** [comment]
@alexwlchan - any second pass comments?  If not I'll merge this soon so that I can start on another tranche without clashes.

File: _RELEASE.rst [hunk 6]
```diff
@@ -29,3 +29,21 @@ Other changes in Hypothesis 4
...
+  in a single call.
+- The deprecated SQLite3 example database has been removed.
+- The ``categories`` arguments to :func:`~hypothesis.strategies.characters`
+  are now strictly validated, so passing a non-existent category is an error.
+- The ``settings`` keyword to `~hypothesis.settings.register_profile` was
+  deprecated in favor of ``parent``, and has been removed.
+- The ``strict`` setting was deprecated (and thus enabling it was an error),
+  and has been removed.
+- Unused Hypothesis exception types have been removed.  (``BadTemplateDraw``,
+  ``DefinitelyNoSuchExample``, ``WrongFormat``, ``BadData``)
```
Comment (alexwlchan) [inline _RELEASE.rst]: nit: missing full stop

File: src/hypothesis/internal/escalation.py [hunk 7]
```diff
@@ -74,9 +74,7 @@ def escalate_hypothesis_internal_error():
         raise
     import traceback
     filepath = traceback.extract_tb(tb)[-1][0]
-    if is_hypothesis_file(filepath) and not isinstance(
-        e, (HypothesisException,) + HYPOTHESIS_CONTROL_EXCEPTIONS,
-    ):
+    if is_hypothesis_file(filepath) and not isinstance(e, HypothesisException):
```
Comment (alexwlchan) [inline src/hypothesis/internal/escalation.py]: Possibly I’ve missed something – why do we no longer exclude the control exceptions? StopTest doesn’t subclass from HypothesisException.

Comment (Zac-HD) [reply src/hypothesis/internal/escalation.py]: Ugh, rebasing error :cry:

**Pull request merged**